job_service/repository/job_db.py

```diff
@@ -65,65 +83,71 @@ def new_job(self, command: str, status: str,
+        if log is not None:
+            job["logs"].append({"at": now, "message": log})
+
+        if command == "ADD_OR_CHANGE_DATA":
             # Atomic insert of new import job
             # Only inserts if no job is found for same datasetname
             # with a currently active status
-            update_result: UpdateResult = self.db.jobs.update_one(
-                {
-                    "datasetName": dataset_name,
-                    "status": {"$nin": ["done", "failed"]}
-                },
-                {"$setOnInsert": job},
-                upsert=True
-            )
+            update_result = None
+            try:
+                update_result: UpdateResult = self.in_progress.update_one(
+                    {"datasetName": dataset_name},
+                    {"$setOnInsert": job},
+                    upsert=True
+                )
+            except DuplicateKeyError:
+                raise JobExistsException(f"{dataset_name} already in progress")
 
             if update_result.upserted_id is None:
                 # Raise exception if active job already existed
-                raise JobExistsException(f"{dataset_name} already imported")
+                raise JobExistsException(f"{dataset_name} already in progress")
             else:
                 return job_id
         else:
-            self.db.jobs.insert_one(job)
+            self.completed.insert_one(job)
             return job_id
 
     def update_job(self, job_id: str, status: str, log: str = None) -> None:
         """
         Updates job with supplied job_id with new status.
         Appends additional log if supplied.
         """
-        self.db.jobs.update_one(
-            {"jobId": job_id},
-            {"$set": {"status": status}}
-        )
         now = datetime.now()
-        self.db.jobs.update_one(
-            {
-                "jobId": job_id
-            },
-            {
-                "$push": {
-                    "log": {"at": now, "message": f"Set status: {status}"}
-                }
+        find_query = {"jobId": job_id}
+        update_status_query = {
+            "$set": {"status": status},
+            "$push": {
+                "log": {"at": now, "message": f"Set status: {status}"}
             }
-        )
+        }
+        add_log_query = {
+            "$push": {
+                "logs": {"at": now, "message": log}
+            }
+        }
+        job = self.in_progress.find_one(find_query)
+        if job is None:
+            raise NotFoundException(f"Could not find job with id {job_id}")
+
+        if status in ["done", "failed"]:
+            self.in_progress.delete_one(find_query)
+            self.completed.insert_one(job)
+            self.completed.update_one(find_query, update_status_query)
+            if log is not None:
+                self.completed.update_one(find_query, add_log_query)
+            return
+
+        self.in_progress.update_one(find_query, update_status_query)
```

Comment:
Updating `in_progress` list when the job might have been deleted with `delete_one`?

Reply:
if it enters the if statement that moves it from inprogress to completed the function terminates on the return on line 145. I can make this an if/else instead if that makes it more explicit?

Reply:
It's OK, I didn't see the return statement :)